receiver/vcenterreceiver/client.go

```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+}
+
+// VSANMetricDetails contains vSAN metric data for a single metric
+type VSANMetricDetails struct {
+	// Contains the metric label
+	MetricLabel string
+	// Contains the metric interval in seconds
+	Interval int32
+	// Contains timestamps for all metric values
+	Timestamps []*time.Time
+	// Contains all values for vSAN metric label
+	Values []int64
+}
+
+// VSANQueryType represents the type of VSAN query
+type VSANQueryType string
+
+const (
+	VSANQueryTypeHosts           VSANQueryType = "host-domclient:*"
+	VSANQueryTypeVirtualMachines VSANQueryType = "virtual-machine:*"
+)
+
+// getLabelsForQueryType returns the appropriate labels for each query type
+func (vc *vcenterClient) getLabelsForQueryType(queryType VSANQueryType) []string {
+	switch queryType {
+	case VSANQueryTypeHosts:
+		return []string{
+			"iopsRead", "iopsWrite", "throughputRead", "throughputWrite",
+			"latencyRead", "latencyWrite", "congestion", "clientCacheHitRate",
+		}
+	case VSANQueryTypeVirtualMachines:
+		return []string{
+			"iopsRead", "iopsWrite", "throughputRead", "throughputWrite",
+			"latencyRead", "latencyWrite",
+		}
+	default:
+		return []string{}
+	}
+}
+
+// VSANQuery performs a VSAN query for the specified type across multiple clusters
+func (vc *vcenterClient) VSANQuery(
+	ctx context.Context,
+	queryType VSANQueryType,
+	clusterRefs []*vt.ManagedObjectReference,
+) (*VSANQueryResults, error) {
+	allResults := VSANQueryResults{
+		MetricResultsByUUID: map[string]*VSANMetricResults{},
+	}
+
+	if vc.vsanDriver == nil {
+		return &allResults, nil
+	}
+
+	for _, clusterRef := range clusterRefs {
+		results, err := vc.vsanQueryByCluster(ctx, queryType, clusterRef)
+		if err != nil {
+			vsanErr := vc.handleVSANError(err, queryType)
+			if vsanErr != nil {
+				return nil, vsanErr
+			}
+			continue
+		}
+
+		maps.Copy(allResults.MetricResultsByUUID, results.MetricResultsByUUID)
+	}
+
+	return &allResults, nil
+}
+
+// vsanQueryByCluster performs a VSAN query for a specific cluster
+func (vc *vcenterClient) vsanQueryByCluster(
+	ctx context.Context,
+	queryType VSANQueryType,
+	clusterRef *vt.ManagedObjectReference,
+) (*VSANQueryResults, error) {
+	queryResults := VSANQueryResults{
+		MetricResultsByUUID: map[string]*VSANMetricResults{},
+	}
+
+	now := time.Now()
+	querySpec := []types.VsanPerfQuerySpec{
+		{
+			EntityRefId: string(queryType),
+			StartTime:   &now,
+			EndTime:     &now,
+			Labels:      vc.getLabelsForQueryType(queryType),
+		},
+	}
+
+	rawResults, err := vc.vsanDriver.VsanPerfQueryPerf(ctx, clusterRef, querySpec)
+	if err != nil {
+		return nil, fmt.Errorf("problem retrieving %s VSAN metrics for cluster %s: %w", queryType, clusterRef.Value, err)
+	}
+
+	for _, rawResult := range rawResults {
+		metricResults, err := vc.convertVSANResultToMetricResults(rawResult)
+		if err != nil {
+			return &queryResults, fmt.Errorf("problem processing %s VSAN metrics for cluster %s: %w", queryType, clusterRef.Value, err)
```

Comment:
vSAN

Reply:
Also, this does simplify things from also checking for `metricResults == nil`, but you lose the ability to check for a specific problem UUID

Reply:
Do you think the simplicity warrants the lack of extra information?

Reply:
Extra information when a customer has a problem is often great to have.